libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/generator/OpenHouseDataLayoutGenerator.java

```diff
@@ -0,0 +1,113 @@
+package com.linkedin.openhouse.datalayout.generator;
+
+import com.linkedin.openhouse.datalayout.config.DataCompactionConfig;
+import com.linkedin.openhouse.datalayout.datasource.FileStat;
+import com.linkedin.openhouse.datalayout.datasource.TableFileStats;
+import com.linkedin.openhouse.datalayout.datasource.TablePartitionStats;
+import com.linkedin.openhouse.datalayout.strategy.DataLayoutOptimizationStrategy;
+import java.util.Collections;
+import java.util.List;
+import lombok.Builder;
+import org.apache.spark.api.java.function.FilterFunction;
+import org.apache.spark.api.java.function.MapFunction;
+import org.apache.spark.sql.Encoders;
+
+/**
+ * Data layout optimization strategies generator for OpenHouse. Generates a list of strategies with
+ * cost.
+ */
+@Builder
+public class OpenHouseDataLayoutGenerator implements DataLayoutGenerator {
+  private static final long MB = 1024 * 1024;
+  private static final long FILE_BLOCK_SIZE_BYTES = 256 * MB;
+  private static final long FILE_BLOCK_MARGIN_BYTES = 10 * MB;
+  private static final long NUM_FILE_GROUPS_PER_COMMIT = 100;
+  private static final long MAX_NUM_COMMITS = 30;
+  private static final long MAX_BYTES_SIZE_RATIO = 10;
+  private static final long REWRITE_BYTES_PER_SECOND = 2 * MB;
+  private static final long EXECUTOR_MEMORY_MB = 2048;
+  private static final int MAX_CONCURRENT_FILE_GROUP_REWRITES = 50;
+  private static final int REWRITE_PARALLELISM = 900; // number of Spark tasks to run in parallel
+  private static final long TARGET_BYTES_SIZE = 2 * FILE_BLOCK_SIZE_BYTES - FILE_BLOCK_MARGIN_BYTES;
+  private static final double COMPUTE_STARTUP_COST_GB_HR = 0.5;
+  private final TableFileStats tableFileStats;
+  private final TablePartitionStats tablePartitionStats;
+
+  /**
+   * Generate a list of data layout optimization strategies based on the table file stats and
+   * historic query patterns.
+   */
+  @Override
+  public List<DataLayoutOptimizationStrategy> generate() {
+    // skip single partition and non-partitioned tables
+    if (tablePartitionStats.get().count() <= 1) {
+      return Collections.emptyList();
+    }
+    return Collections.singletonList(generateCompactionStrategy());
+  }
+
+  private DataLayoutOptimizationStrategy generateCompactionStrategy() {
+    List<Long> fileSizes =
+        tableFileStats
+            .get()
+            .map((MapFunction<FileStat, Long>) FileStat::getSize, Encoders.LONG())
+            .filter(
+                (FilterFunction<Long>)
+                    size ->
+                        size < TARGET_BYTES_SIZE * DataCompactionConfig.MIN_BYTE_SIZE_RATIO_DEFAULT)
+            .collectAsList();
```

Comment:
We are getting all the fileSizes in-memory and then computing sum() and count(). Does it make sense for us to delegate the computation for sum of fileSizes that are candidate for compaction to spark?
Cache it after computing it once using spark to be used later.

Reply:
I didn't want to optimize prematurely, 8 bytes per file * 10m = 80MiB. After giving it more thought and experiencing recent memory tuning, I think it's worth computing in Spark, it's fast.

Reply:
Yes I also have recency bias when I made that comment. :)

Reply:
FIxed.